File: nodejs/secure_channel.js
const crypto = require('crypto')
const eckey = require('ecdsa-secp256r1')
const bigint = require('bigi')
const ecurve = require('ecurve')

const ec256r1 = ecurve.getCurveByName('secp256r1')
const ecdh = crypto.createECDH('prime256v1') // secp256r1 = prime256v1

function generateECKeyPair() {
    ecdh.generateKeys()

    skBArr = ecdh.getPrivateKey() // 32 bytes
    pkBArr = ecdh.getPublicKey() // 65 bytes

    for (let i = skBArr.length; i < 32; i++) {
        skBArr = Buffer.concat([Buffer.from('00', 'hex'), skBArr])
    }

    skHex = skBArr.toString('hex')
    pkHex = pkBArr.slice(1).toString('hex')

    // curve prime256v1
    const ecKeyPair = {
        PublicKey: pkHex, // format x || y, 64 bytes in hex string
        PrivateKey: skHex // 32 bytes in hex string
    }

    return ecKeyPair
}

function calSignedData(selfSkHexStr, data) {

    let dataByteArr = Buffer.from(data)
    let hash = crypto.createHash('sha256').update(dataByteArr).digest().toString('hex')

    let sign = crypto.createSign('sha256');
    sign.update(dataByteArr) // byte array
    sign.end()

    let skPem = recoverPemPrivateKey(selfSkHexStr)
    let sig = sign.sign({ key: skPem, dsaEncoding: 'ieee-p1363' }).toString('hex')

    const signedData = {
        Data: data,
        Hash: hash, // hash of data, length 32 bytes in hex string
        Sig: sig // ieee-p1363 format, a.k.a r||s, length 64 bytes in hex string
    }
    
    return signedData
}

function verifySignedData(fromPkHexStr, signedData) {

    let dataByteArr = Buffer.from(signedData.Data)
    let hashCal = crypto.createHash('sha256').update(dataByteArr).digest().toString('hex')

	// verify hash
	if (hashCal != signedData.Hash) {
        return false
	}

    // verify signature

    if(signedData.Sig.length != 128) {
        throw new TypeError("invalid signature, should in ieee-p1363 format, a.k.a. r || s with length 64 bytes")
	}

    let sigByteArr = Buffer.from(signedData.Sig, 'hex')
    let pkPem = recoverPemPublicKey(fromPkHexStr)

    let  verify = crypto.createVerify('sha256');
    verify.update(dataByteArr);
    verify.end();

    let result = verify.verify( { key: pkPem, dsaEncoding: 'ieee-p1363'}, sigByteArr)

    return result
}

function calEncSignedData(selfSkHexStr, toPkHexStr, data) {

    let signedData = calSignedData(selfSkHexStr, data)

    let r = crypto.randomBytes(32);
    let iv = crypto.randomBytes(16);
    let ivHexStr = iv.toString('hex')

    let rBigI = bigint.fromBuffer(r)
    rBigI = rBigI.mod(ec256r1.n)
    let RPoint = ec256r1.G.multiply(rBigI)
    let RxHexStr = RPoint.affineX.toBuffer(32).toString('hex')
    let RyHexStr = RPoint.affineY.toBuffer(32).toString('hex')

    let pkXBigI = bigint.fromBuffer(Buffer.from(toPkHexStr.slice(0, 64), 'hex'))
    let pkYBigI = bigint.fromBuffer(Buffer.from(toPkHexStr.slice(64), 'hex'))

    let PKPoint = ecurve.Point.fromAffine(ec256r1, pkXBigI, pkYBigI)
    let RPKPoint = PKPoint.multiply(rBigI)

    // AES key: hash(rPKx), (rPKx, rPKy) = r * pk = sk * R

    let keyAes = crypto.createHash('sha256').update(RPKPoint.affineX.toBuffer(32)).digest()

	// AES CTR encrypt

    let encrypt = crypto.createCipheriv('aes-256-ctr', keyAes, iv);
    let ciphertext = encrypt.update(Buffer.from(data));
    let ciphertextHexStr = Buffer.concat([ciphertext, encrypt.final()]).toString('hex')

    const encSignedData = {
        DataEnc: ivHexStr + RxHexStr + RyHexStr + ciphertextHexStr, 
        Hash: signedData.Hash, // hash of data, length 32 bytes in hex string
        Sig: signedData.Sig // ieee-p1363 format, a.k.a r||s, length 64 bytes in hex string
    }
    
    return encSignedData
}

function verifyEncSignedData(selfSkHexStr, fromPkHexStr, encSignedData) {

    let ivHexStr = encSignedData.DataEnc.slice(0,32)
    let RxHexStr = encSignedData.DataEnc.slice(32,96)
    let RyHexStr = encSignedData.DataEnc.slice(96,160)
    let ciphertextHexStr = encSignedData.DataEnc.slice(160)

    let iv = Buffer.from(ivHexStr, 'hex')

    let RxBigI = bigint.fromBuffer(Buffer.from(RxHexStr, 'hex'))
    let RyBigI = bigint.fromBuffer(Buffer.from(RyHexStr, 'hex'))
    let skBigI = bigint.fromBuffer(Buffer.from(selfSkHexStr, 'hex'))

    let RPoint = ecurve.Point.fromAffine(ec256r1, RxBigI, RyBigI)
    let RPKPoint = RPoint.multiply(skBigI)

    // AES key: hash(rPKx), (rPKx, rPKy) = r * pk = sk * R

    let keyAes = crypto.createHash('sha256').update(RPKPoint.affineX.toBuffer(32)).digest()

	// AES CTR decrypt
	// note: decrypt always success, however is not the real plaintext with wrong selfSkHexStr

    let decrypt = crypto.createDecipheriv('aes-256-ctr', keyAes, iv);
    let plaintextBytes = decrypt.update(Buffer.from(ciphertextHexStr, 'hex'));
    plaintextBytes = Buffer.concat([plaintextBytes, decrypt.final()]);

    let plaintext = plaintextBytes.toString()
    
	// verify hash and signature

	let signedData = {
        Data: plaintext,
        Hash: encSignedData.Hash,
        Sig: encSignedData.Sig
    }

	return verifySignedData(fromPkHexStr, signedData)
}

function recoverPemPrivateKey(skHexStr) {
    if (skHexStr.length != 64 || skHexStr.length %2 != 0){
        throw new TypeError('invalid EC private key, wrong length')
    }

    let skByteArr = Buffer.from(skHexStr, 'hex');
    let skBigI = bigint.fromBuffer(skByteArr)
    var pkPoint = ec256r1.G.multiply(skBigI)

    let sk = new eckey({
        d: skByteArr,
        x: pkPoint.affineX.toBuffer(32),
        y: pkPoint.affineY.toBuffer(32)
    })

    return sk.toPEM()
}

function recoverPemPublicKey(pkHexStr) {
    if (pkHexStr.length != 128 || pkHexStr.length %2 != 0){
        throw new TypeError('invalid EC point public key, wrong length')
    }

    let pkByteArr = Buffer.from(pkHexStr, 'hex');

    let pk = new eckey({
        x: pkByteArr.slice(0, 32),
        y: pkByteArr.slice(32)
    })

    return pk.toPEM()
}

module.exports = {
    generateECKeyPair: generateECKeyPair,
    calSignedData: calSignedData,
    verifySignedData: verifySignedData,
    calEncSignedData: calEncSignedData,
    verifyEncSignedData: verifyEncSignedData
};

File: golang/secure_channel/secure_channel.go
package secure_channel

import (
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"errors"
	"math/big"
	"io"
	"crypto/sha256"
	"crypto/aes"
	"crypto/cipher"
)

type SignedData struct {
	Data string
	Hash string // hash of data, length 32 bytes in hex string
	Sig	string // ieee-p1363 format, a.k.a r||s, length 64 bytes in hex string
	IdFrom string // optional, identifier of sender
}

type EncSignedData struct {
	DataEnc string // IV (16 bytes), R (64 bytes) and ciphertext of data in hex string, encrypted by ecies (AES CTR mode)
	Hash string // hash of data, length 32 bytes in hex string
	Sig	string // ieee-p1363 format, a.k.a r||s, length 64 bytes in hex string
	IdFrom string // optional, identifier of sender
}

// curve prime256v1
type ECKeyPair struct{
	PublicKey string // format x || y, 64 bytes in hex string
	PrivateKey string // 32 bytes in hex string
}

func GenerateECKeyPair() (ecKeyPair *ECKeyPair, err error) {

	keyPair, errKeyGen := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	if errKeyGen != nil {
		return nil, errKeyGen
	}

	xStr, errXStr := BigInt2HexString(keyPair.X, 32)
	yStr, errYStr := BigInt2HexString(keyPair.Y, 32)

	skStr, errSkStr := BigInt2HexString(keyPair.D, 32)

	if errXStr != nil || errYStr != nil || errSkStr != nil {
		return nil, errors.New("failed to format the EC keypair to hex string")
	}

	ecKeyPair = &ECKeyPair{
		PublicKey: xStr + yStr,
		PrivateKey: skStr,
	}

	return
}

func CalSignedData(selfSkHexStr string, data string) (signedData *SignedData, err error) {

	hash := sha256.Sum256([]byte(data))
	hashHexStr := BytesArray2HexString(hash[:])

	sk, errSk := recoverPrivateKey(selfSkHexStr)
	if errSk != nil {
		return nil, errSk
	}

	r, s, errSig := ecdsa.Sign(rand.Reader, sk, hash[:])
	if errSig != nil {
		return nil, errors.New("failed to compute the signature, " + errSig.Error())
	}

	rStr, errR := BigInt2HexString(r, 32)
	sStr, errS := BigInt2HexString(s, 32)

	if errR != nil || errS != nil {
		return nil, errors.New("failed to format the signature to hex string")
	}
	
	signedData = &SignedData{
		Data: data,
		Hash: hashHexStr,
		Sig: rStr + sStr,
	}
	
	return
}

func VerifySignedData(fromPkHexStr string, signedData *SignedData) (data string, err error) {

	hashCal := sha256.Sum256([]byte(signedData.Data))
	hashCalHexStr := BytesArray2HexString(hashCal[:])

	// verify hash

	if hashCalHexStr != signedData.Hash {
		return "", errors.New("invalid data, NOT pass hash validation")
	}

	pk, errPk := recoverPublicKey(fromPkHexStr)
	if errPk != nil {
		return "", errPk
	}

	if len(signedData.Sig) != 128 {
		return "", errors.New("invalid signature, should in ieee-p1363 format, a.k.a. r || s with length 64 bytes")
	}

	rBigInt, errR := HexString2BigInt(signedData.Sig[:64])
	sBigInt, errS := HexString2BigInt(signedData.Sig[64:])

	if errR != nil || errS != nil {
		return "", errors.New("invalid signature, failed to convert the hex string to big.Int")
	}

	// verify signature

	result := ecdsa.Verify(pk, hashCal[:], rBigInt, sBigInt)
	if !result {
		return "", errors.New("invalid signature, NOT pass signature validation")
	}

	return signedData.Data, nil
}

func CalEncSignedData(selfSkHexStr string, toPkHexStr string, data string) (encSignedData *EncSignedData, err error) {

	signedData, errSigned := CalSignedData(selfSkHexStr, data)
	if errSigned != nil {
		return nil, errSigned
	}

	pk, errPk := recoverPublicKey(toPkHexStr)
	if errPk != nil {
		return nil, errPk
	}

	r, errR := randomInt(32)
	iv, errIv := randomBytes(16)

	if errR != nil || errIv != nil {
		return nil, errors.New("failed to get random big.Int")
	}

	ivHexStr := BytesArray2HexString(iv)

	r = new(big.Int).Mod(r, elliptic.P256().Params().N)
	rPKx, _ := elliptic.P256().ScalarMult(pk.X, pk.Y, r.Bytes())
	Rx, Ry := elliptic.P256().ScalarBaseMult(r.Bytes())

	RxHexStr, errRx := BigInt2HexString(Rx, 32)
	RyHexStr, errRy := BigInt2HexString(Ry, 32)

	if errRx != nil || errRy != nil {
		return nil, errors.New("failed to format the R point to hex string")
	}

	// AES key: hash(rPKx), (rPKx, rPKy) = r * pk = sk * R

	keyAes := sha256.Sum256(rPKx.Bytes())

	// AES CTR encrypt

	aesBlock, errAes := aes.NewCipher(keyAes[:])
	if errAes != nil {
		return nil, errors.New("failed to initialize the AES encryption")
	}

	ciphertextBytes := make([]byte, len([]byte(data)))

	stream := cipher.NewCTR(aesBlock, iv)
	stream.XORKeyStream(ciphertextBytes, []byte(data))

	ciphertextHexStr := BytesArray2HexString(ciphertextBytes)

	encSignedData = &EncSignedData{
		DataEnc: ivHexStr + RxHexStr + RyHexStr + ciphertextHexStr,
		Hash: signedData.Hash,
		Sig: signedData.Sig,
	}

	return
}

func VerifyEncSignedData(selfSkHexStr string, fromPkHexStr string, encSignedData *EncSignedData) (data string, err error) {

	iv, errIv := HexString2BytesArray(encSignedData.DataEnc[:32])
	if errIv != nil {
		return "", errors.New("failed to get IV from encrypted data")
	}

	Rx, errRx := HexString2BigInt(encSignedData.DataEnc[32:96])
	Ry, errRy := HexString2BigInt(encSignedData.DataEnc[96:160])

	if errRx != nil || errRy != nil {
		return "", errors.New("failed to format hex string to R point")
	}

	skBitInt, errSk := HexString2BigInt(selfSkHexStr)
	if errSk != nil {
		return "", errors.New("failed to format the sk hex string to big.Int, " + errSk.Error())
	}

	rPKx, _ := elliptic.P256().ScalarMult(Rx, Ry, skBitInt.Bytes())

	// AES key: hash(x), (x, y) = r * pk = sk * R
	
	keyAes := sha256.Sum256(rPKx.Bytes())

	// AES CTR decrypt
	// note: decrypt always success, however is not the real plaintext with wrong selfSkHexStr

	aesBlock, errAes := aes.NewCipher(keyAes[:])
	if errAes != nil {
		return "", errors.New("failed to initialize the AES decryption")
	}

	ciphertextBytes, errCipher := HexString2BytesArray(encSignedData.DataEnc[160:])
	if errCipher != nil {
		return "", errors.New("failed to format the cipher in hex string to byte array, " + errCipher.Error())
	}

	plaintextBytes := make([]byte, len(ciphertextBytes))

	stream := cipher.NewCTR(aesBlock, iv)
	stream.XORKeyStream(plaintextBytes, ciphertextBytes)

	dataDecrypt := string(plaintextBytes[:])

	// verify hash and signature

	signedData := &SignedData{
		Data: dataDecrypt,
		Hash: encSignedData.Hash,
		Sig: encSignedData.Sig,
	}

	return VerifySignedData(fromPkHexStr, signedData)
}


func recoverPrivateKey(hexStr string) (*ecdsa.PrivateKey, error) {

	skBigInt, errSk := HexString2BigInt(hexStr)

	if errSk != nil {
		return nil, errors.New("failed to get private key from hex string, " + errSk.Error())
	}

	sk := new(ecdsa.PrivateKey)
	sk.PublicKey.Curve = elliptic.P256()
	sk.D = skBigInt
	sk.PublicKey.X, sk.PublicKey.Y = elliptic.P256().ScalarBaseMult(skBigInt.Bytes())

	return sk, nil
}

func recoverPublicKey(hexStr string) (*ecdsa.PublicKey, error) {

	if len(hexStr) != 128 {
		return nil, errors.New("failed to get public key from hex string, invalid public key hex string (x || y format), length should be 64 bytes")
	}

	XBigInt, errX := HexString2BigInt(hexStr[:64])
	YBigInt, errY := HexString2BigInt(hexStr[64:])

	if errX != nil || errY != nil {
		return nil, errors.New("failed to get public key from hex string")
	}

	pk := new(ecdsa.PublicKey)
	pk.Curve = elliptic.P256()
	pk.X, pk.Y = XBigInt, YBigInt

	return pk, nil
}

func randomInt(byteLen int) (*big.Int, error) {

	b := make([]byte, byteLen)
	
	_, err := io.ReadFull(rand.Reader, b)
	if err != nil {
		return nil, err
	}

	k := new(big.Int).SetBytes(b)

	return k, nil
}

func randomBytes(byteLen int) ([]byte, error) {

	b := make([]byte, byteLen)
	
	_, err := io.ReadFull(rand.Reader, b)
	if err != nil {
		return nil, err
	}

	return b, nil
}
